### DISCUSSION TITLE
Mypy and pyright don't agree about variance for Generic Protocol?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#5212] [Q&A]
```python
from typing import Generic, Iterable, Iterator, Protocol, TypeVar, runtime_checkable

T = TypeVar("T")

@runtime_checkable
class _ConstructableIterable(Generic[T], Protocol):  # mypy : Invariant type variable "T" used in protocol where covariant one is expected
    def __init__(self, __iterable: Iterable[T]) -> None:
        ...

    def __iter__(self) -> Iterator[T]:
        ...


T_co = TypeVar("T_co", covariant=True)

@runtime_checkable
class _ConstructableIterable_co(Generic[T_co], Protocol):  # pyright : Type variable "T_co" used in generic protocol "_ConstructableIterable_co" should be invariant
    def __init__(self, __iterable: Iterable[T_co]) -> None:
        ...

    def __iter__(self) -> Iterator[T_co]:
        ...
```

I feel like I am inclined to believe that mypy is correct here because `typing.Iterable` is typed as being covariant: https://python.readthedocs.io/en/stable/library/typing.html?highlight=str#typing.Iterable.

Pylance `v2023.5.40` for VS Code

**Comment by erictraut** [comment]
This appears to be a bug in mypy. The variance of `T` is not related to the variance of the type parameter used for `typing.Iterable`. As evidence of this, you can replace `Iterable[T]` with `T` and see the same behavior.

The variance of `T` should be determined by whether it's used in input parameters, return values, or both. In this case, it's being used in both (an input parameter in `__init__` and a return value for `__iter__`). It appears that mypy is skipping the `__init__` method for purposes of variance calculation, but I don't see any justification for doing so. If I change the name of the `__init__` to something else (e.g. `foo`), then mypy correctly determines that `T` should be invariant.

**Reply by ringohoffman** [reply]
I'm not sure I understand variance well enough to judge this response, but here it is: 
 https://github.com/python/mypy/issues/15331#issuecomment-1583610642

**Reply by ringohoffman** [reply]
BTW, to clarify your explanation... is it

* Invariant for both input and outputs
* covariant for inputs only
* contravariant for outputs only

?

Also do you have any resource you can link for this? I'd never heard of this method of determining variance.

**Reply by erictraut** [reply]
I responded to the thread in the mypy tracker. I've never seen any documentation that mypy skips `__init__` when computing variance, nor have I heard a clear, principled reason for doing so. If you're interested in the details of variance calculation, you may find [this section of PEP 695](https://peps.python.org/pep-0695/#variance-inference) useful. This is the technique that both pyright and mypy (and presumably other Python type checkers) use for determining variance of type variables. Note that I'm the primary author of PEP 695.

**Reply by hmc-cs-mdrissi** [reply]
This reminds me of this issue, https://github.com/microsoft/pyright/issues/3839

Pyright also partly skips TypeVar in init and new at least for case of,

```python
T = TypeVar(‘T’, covariant=True)

class Box(Generic[T])
  def __init__(self, x: T) -> None:
    …
```

type checks fine even though x is used as input argument. I think mypy’s argument is same as why that type checks even though changing function name to any other doesn’t.

I don’t even see what is unsafe about allowing T as covariant here for normal usage of specifying Box[T] as an annotation. I can see type[Box[T]] being hole, but pragmatically necessary for most generic types.

Edit: I think underlying principle why init and new are special is they are methods normally only called on class and not instance of that class. You don’t really see code like,

```python
x = Bar(4)
x.__init__(3) # unsure what this would even do besides crash.
```

edit 2: One strange thing about original example is `__init__` with a protocol. That I've never seen especially with `__init__` tends to be exempted for override compatibility. 

This code passes pyright,

```python
T = TypeVar('T', covariant=True)

class Box(Generic[T])
  def __init__(self, x: T) -> None:
    …
```

while this doesn't

```python
T = TypeVar('T', covariant=True)

class Box(Protocol[T])
  def __init__(self, x: T) -> None:
    …
```

So rule here seems special for protocols and unsure why.

**Reply by erictraut** [reply]
I think you're conflating two different variance-related checks:
1. Is the variance of a class-scoped type variable correct for all usages within that class?
2. Does the type annotation for an input parameter or return type make use of a type variable with an incompatible variance?

The first check is currently performed only on protocol classes (in both mypy and pyright). The second check applies to methods in all classes, whether or not they are protocols. As you've pointed out, `__init__` and `__new__` methods are exempted from the second check. 

This thread is focused on the first issue. I'm especially interested in pinning down the correct behavior here because it matters for variance inference (introduced in PEP 695). Currently, pyright does not exempt `__init__` and `__new__` methods when performing the first check, but mypy apparently does.

I think you make an interesting argument about why `__init__` and `__new__` should be exempt from the first check. I'll think more about this and see if I can come up with some counterexamples. 

@ilevkivskyi, is this the same reasoning you have in mind? Can you think of any typing holes that this could create?
